### DISCUSSION TITLE
[CH5] 파싱이란?

### DISCUSSION THREAD

**Discussion by innerstella** [innerstella/ES2024#4] [Q&A] [answered]
Q : “자바스크립트 엔진이 자바스크립트 코드를 계산(평가)하려면 먼저 기호(리터럴과 연산자)의 의미를 알고 있어야 하며, 표현식의 의미도 해석(파싱)할 수 있어야 한다.”가 시작할때 나오는데 각각 평가, 리터럴과 연산자 , 표현식, 파싱에 대해서.

**Comment by Lv1GoM** [comment, marked as answer]
### 파싱(Parsing)
파싱은 소스 코드를 분석하고 문법 규칙에 따라 ***추상 구문 트리(Abstract Syntax Tree, AST)***를 생성하는 과정을 말한다
자바스크립트 엔진은 코드를 읽어들이고 이를 토큰으로 분해한 다음, 이를 기반으로 추상 구문 트리를 생성하여 실행

### 파싱의 단계
#### 토큰화 (Lexical Analysis)
소스 코드를 토큰(Token)으로 분해하는 단계
토큰은 언어의 기본 구성 단위로, 식별자, 예약어, 연산자, 숫자, 문자열 등이 될 수 있다.
토큰화는 일련의 문자열을 토큰으로 변환하는 작업을 수행한다

#### 구문 분석 (Syntax Analysis)
토큰화된 토큰들을 이용하여 문법 규칙에 따라 문장의 구조를 해석하는 단계
문법 규칙은 언어의 문법에 따라 정의되며, 이를 기반으로 추상 구문 트리(AST)를 생성
일반적으로 구문 분석은 컨텍스트-자유 문법(Context-Free Grammar, CFG)을 기반으로 한다,

#### AST 생성 (Abstract Syntax Tree)
구문 분석의 결과로 추상 구문 트리가 생성된다.
AST는 소스 코드의 추상적인 구조를 나타내며, 트리 구조로 표현된다.
각 노드는 토큰 또는 표현식을 나타내고, 노드 간의 관계는 문법 규칙에 따라 정의된다.

```
let sum = 2 + 3; //소스코드
[let] [sum] [=] [2] [+] [3] [;] //토큰화
AssignmentExpression(할당표현식) //구문 분석 및 AST
├── Identifier (sum) (식별자)
└── BinaryExpression  (이항 표현식)
    ├── Literal (2)
    ├── Operator (+)
    └── Literal (3)
```
